File: src/arch/x86_64/idt.zig
// Arcturus - Hobbyist operating system written in Zig.
// Copyright (C) 2025 Theomund
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

const instruction = @import("instruction.zig");
const isr = @import("isr.zig");

const Descriptor = packed struct {
    offset_1: u16,
    selector: u16,
    interrupt_stack_table: u3,
    reserved_1: u5,
    gate_type: u4,
    reserved_2: u1,
    desired_privilege_level: u2,
    present: u1,
    offset_2: u48,
    reserved_3: u32,
};

pub const Pointer = extern struct {
    limit: u16 align(1),
    base: u64 align(1),
};

pub const Table = struct {
    descriptors: [256]u128,

    pub fn init(selector: u16) Table {
        const null_descriptor = Descriptor{
            .offset_1 = 0x0,
            .selector = selector,
            .interrupt_stack_table = 0,
            .reserved_1 = 0,
            .gate_type = 0xE,
            .reserved_2 = 0,
            .desired_privilege_level = 0,
            .present = 0,
            .offset_2 = 0x0,
            .reserved_3 = 0,
        };

        const breakpoint_offset = @intFromPtr(&isr.breakpoint_exception);

        const breakpoint_descriptor = Descriptor{
            .offset_1 = @truncate(breakpoint_offset),
            .selector = selector,
            .interrupt_stack_table = 0,
            .reserved_1 = 0,
            .gate_type = 0xF,
            .reserved_2 = 0,
            .desired_privilege_level = 0,
            .present = 1,
            .offset_2 = @truncate(breakpoint_offset >> 16),
            .reserved_3 = 0,
        };

        var descriptor_table = [_]u128{@bitCast(null_descriptor)} ** 256;

        descriptor_table[3] = @bitCast(breakpoint_descriptor);

        return Table{
            .descriptors = descriptor_table,
        };
    }

    fn pointer(self: Table) Pointer {
        return Pointer{
            .limit = (self.descriptors.len * @sizeOf(u128)) - 1,
            .base = @intFromPtr(&self.descriptors[0]),
        };
    }

    pub fn load(self: Table) void {
        const ptr = self.pointer();
        instruction.lidt(&ptr);
        instruction.sti();
        instruction.int3();
    }
};

File: kernel/src/idt.rs
// Arcturus - Hobbyist operating system written in Rust.
// Copyright (C) 2025 Theomund
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

use architecture::x86_64::idt::{Handlers, Table};
use core::cell::LazyCell;
use utility::info;
use utility::lock::Spinlock;

use crate::gdt::GDT;
use crate::isr::{
    alignment_check_handler, bound_range_exceeded_handler, breakpoint_handler,
    control_protection_handler, debug_handler, device_not_available_handler,
    division_error_handler, double_fault_handler, general_protection_fault_handler,
    hypervisor_injection_handler, invalid_opcode_handler, invalid_tss_handler,
    machine_check_handler, non_maskable_interrupt_handler, overflow_handler, page_fault_handler,
    security_handler, segment_not_present_handler, simd_floating_point_handler,
    stack_segment_fault_handler, virtualization_handler, vmm_communication_handler,
    x87_floating_point_handler,
};

pub static IDT: Spinlock<LazyCell<Table>> = Spinlock::new(LazyCell::new(|| {
    let handlers = Handlers {
        division_error_handler,
        debug_handler,
        non_maskable_interrupt_handler,
        breakpoint_handler,
        overflow_handler,
        bound_range_exceeded_handler,
        invalid_opcode_handler,
        device_not_available_handler,
        double_fault_handler,
        invalid_tss_handler,
        segment_not_present_handler,
        stack_segment_fault_handler,
        general_protection_fault_handler,
        page_fault_handler,
        x87_floating_point_handler,
        alignment_check_handler,
        machine_check_handler,
        simd_floating_point_handler,
        virtualization_handler,
        control_protection_handler,
        hypervisor_injection_handler,
        vmm_communication_handler,
        security_handler,
    };

    Table::new(&handlers, GDT.lock().selector(1))
}));

pub fn init() {
    IDT.lock().load();
    info!("Initialized the interrupt descriptor table.");
}
